[master] Use unittest.mock for newer Python versions

ycedres opened the pull request:
### What does this PR do?

This PR causes the test suite to use `unittest.mock` instead of `mock` for supported Python versions, so the usage of standard library components is favoured.

To achieve this, the `__version__` variable is removed as it was dropped from Python 3.9  and it's almost not used.

The results of the test suite do not change.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

twangboy reviewed (changes requested):
This needs to be rebased and conflicts resolved.

dwoz committed:
Use unittest.mock instead of external mock package on Python >= 3.8

On Python 3.8+, unittest.mock from the standard library is sufficient.
Fall back to the external mock package only on older Python versions.

Adds AsyncMock to re-exports (was missing in original draft).
Drops mock from static requirements for py3.8+ platforms.
